Introduce pushdownFilter SPI

mbasmanova opened the pull request:
Introduce `pushdownFilter` metadata API to enable more efficient filter processing for ORC data inside of the Hive connector. Also, introduce `isPushdownFilterSupported` API to allow connectors to opt-in into the new functionality. By default, `isPushdownFilterSupported` returns `false` and `pushdownFilter` is not invoked. When `isPushdownFilterSupported` returns true, the engine invokes `pushdownFilter` instead of `getTableLayouts` API. The connector encodes the pushed down filter into the `ConnectorTableLayoutHandle`.

This PR introduces the new APIs, updates `PickTableLayouts` optimizer rule to use them and provides a minimal implementation for the Hive connector.

The implementation in Hive connector covers primarily the optimizer part and there is no logic to evaluate the pushed down filter during the execution yet. The information about the filter is propagated via `HiveSplit` and this PR includes the necessary plumbing. The original code path is modified slightly to compact range filters during the execution instead of planning. This is done to reduce the discrepancy between two code paths in the planning phase.

Aside from `PickTableLayouts` rule, there are a number of places that invoke `getTableLayout` API. These will be updated in subsequent PRs.

highker commented on presto-main/src/main/java/com/facebook/presto/metadata/Metadata.java:
```diff
@@ -94,6 +95,10 @@
      */
     TableHandle getAlternativeTableHandle(Session session, TableHandle tableHandle, PartitioningHandle partitioningHandle);
 
+    boolean isPushdownFilterSupported(Session session, TableHandle tableHandle);
```
When filter node and scan node get moved to SPI, this interface can be replaced with a connector optimizer rule (by moving the predicate in filter node to table scan's table layout). So shall we add a javadoc to these two interfaces to indicate the interfaces will be removed once we have plan node in SPI (with a ticket number 12546) maybe?

wenleix replied:
ditto for `pushdownFilter`

tdcmeehan replied:
We can add it now if we merge #12904 :D

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java:
```diff
@@ -239,6 +273,66 @@ public Result apply(TableScanNode tableScanNode, Captures captures, Context cont
         }
     }
 
+    private static final class TranslateVariableNamesVisitor
```
Actually @hellium01 has a nice util to do the same thing (https://github.com/prestodb/presto/pull/12710/commits/21af13229d4491945a8a7061dd980b3252e5915e) with test coverage. Could you check that out? I have asked him to separate that out as a single PR (#12881)

hellium01 replied:
It might change in the near future as @rongrong is replacing all `Symbol` with `VariableExpression`.

rongrong replied:
You have `TypeProvider` here, so you can actually construct a proper `VariableReferenceExpressionInliner` It's actually not safe to use the reference variable's type as the inlined variable's type (I've seen test failures related to this). So the proper way to do it is to use `TypeProvider` to construct `VariableReferenceExpression` and use the variable mapping as context.

hellium01 replied:
I think what you actually need is something like `RowExpressionNodeInliner` to replace `VariableReferenceExpression` with `ColumnReferenceExpression` back/forth?

mbasmanova replied:
@rongrong @highker @hellium01 Yi, I found `RowExpressionNodeInliner`. It allows one to replace one RowExpression with another. I could use it to replace VariableReferenceExpression with VariableReferenceExpression, but I don't see ColumnReferenceExpression. Are you suggesting to introduce one?

hellium01 replied:
I thought we already introduced `ColumnReferenceExpression` that uses ColumnHandle. If not, can you add support for that? It is going to be very useful for pushdown cases.

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java:
```diff
@@ -122,6 +124,22 @@ default ConnectorTableLayoutHandle getAlternativeLayoutHandle(ConnectorSession s
+    /**
+     * Experimental: if true, the engine will invoke pushdownFilter instead of getTableLayouts.
+     */
+    default boolean isPushdownFilterSupported(ConnectorSession session, ConnectorTableHandle tableHandle)
```
Maybe add a comment saying that it will be replaced once `FilterNode` is in SPI (#12546)

wenleix replied:
Curious: Why this SPI needs to have `tableHandle`? 

- For now it returns true for partitioned table, and false for unpartitioned table. However the predicate being pushed down might not include partition columns. So this check doesn't look to me to be necessary? 
- In the future we can pushdown complex type extractions, so any table with complex type is candidate for pushdown filter.

mbasmanova replied:
@wenleix In this iteration of Aria, filter pushdown is supported only for tables stored using ORC or DWRF formats. In the future, we hope to add support for Parquet. Technically speaking, I could move that check into the pushdownFilter method itself, but it seems cleaner and safer to have it here.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java:
```diff
@@ -213,13 +230,30 @@ public boolean isEnabled(Session session)
 
+            Session session = context.getSession();
+            if (metadata.isPushdownFilterSupported(session, tableHandle)) {
+                PushdownFilterResult pushdownFilterResult = metadata.pushdownFilter(session, tableHandle, TRUE);
```
If we push `TRUE` into table scan, we are not expecting taking back any unenforced constraint of `isNone()` case right? Or I might have missed something.

hellium01 replied:
Yes, it seems like should be the case.

mbasmanova replied:
@highker @hellium01 This is existing logic. If the table is empty and the connector has a way to detect this with certainty, it will return `none`. Here is the corresponding snippet of code for the original code path.

```
            TableLayoutResult layout = metadata.getLayout(
                    session,
                    tableHandle,
                    Constraint.alwaysTrue(),
                    Optional.of(tableScanNode.getOutputSymbols().stream()
                            .map(tableScanNode.getAssignments()::get)
                            .collect(toImmutableSet())));

            if (layout.getLayout().getPredicate().isNone()) {
                return Result.ofPlanNode(new ValuesNode(context.getIdAllocator().getNextId(), tableScanNode.getOutputSymbols(), ImmutableList.of()));
            }
```

wenleix replied:
Correct me if I am wrong, but looks to me line 234-245 is sort of doing the same thing as line 248-266, but using the `pushdownFilter` with `TRUE` instead of using `getLayout` with `Constraint.alwaysTrue()`.

I am curious if we can just use the original logic form line 248-266? Since my understanding is `metadata .getLayout` is a more universal method and implemented for every connector. And `pushdownFilter` method is more suitable for `pushPredicateIntoTableScan`.

Or, is the idea here that for a connector supports `pushdownFilter`, this `pushdownFilter` supersedes `getLayout` for consistency purpose?

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java:
```diff
@@ -250,9 +344,49 @@ public static PlanNode pushPredicateIntoTableScan(
+                    .collect(toImmutableBiMap(
+                            entry -> entry.getKey().getName(),
+                            entry -> getColumnName(session, metadata, node.getTable(), entry.getValue())));
+            RowExpression translatedPredicate = SqlToRowExpressionTranslator.translate(predicate, predicateTypes, ImmutableMap.of(), metadata.getFunctionManager(), metadata.getTypeManager(), session, false)
```
Let's see if this can be altered based on @rongrong's comment (https://github.com/prestodb/presto/pull/12875/commits/79b83209b10feaffd8c0cb63b7a51fdda47df160#r289458788).

hellium01 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java:
```diff
@@ -250,9 +344,49 @@ public static PlanNode pushPredicateIntoTableScan(
+            RowExpression translatedPredicate = SqlToRowExpressionTranslator.translate(predicate, predicateTypes, ImmutableMap.of(), metadata.getFunctionManager(), metadata.getTypeManager(), session, false)
+                    .accept(new TranslateVariableNamesVisitor(), symbolToColumnNameMap);
+
+            PushdownFilterResult pushdownFilterResult = metadata.pushdownFilter(session, node.getTable(), translatedPredicate);
```
I haven't gone into HiveConnector's code but it looks like it will replace the old `pickTableLayout`. There is a few things likely needs to be careful. One thing is that there is a few hacks in push down predicate on partition keys to do some evaluation on complicated predicate and not return huge tupleDomain. That's why constraint is passed in with a  `isCandidate`. 

Perhaps, that is something who knows HiveConnector can chime in.

hellium01 replied:
Another place might need to change is `MetadataQueryOptimizer`. Otherwise, we will has to keep both methods in `HiveConnector`.

mbasmanova replied:
@hellium01 Yi, indeed, there are a number of places that invoke Metadata.getLayout and all of these will be need to be modified to use pushdownFilter as well. One option would be to fold filterPushdown into Metadata.getLayout. I plan on working through this in a separate PR, because I feel this PR is pretty big already. I would also like to set things up so that multiple people could work in parallel, e.g. once this PR lands, one person can work on the above while another one can work on subfield pruning and/or implementing filter evaluation in Hive connector.

Re: isCandidate - My understanding is that the new code path (via pushdownFilter) is missing out on some opportunities to prune partitions using deterministic expressions, e.g. `SELECT * FROM t WHERE length(p) > 10`. That's because, `HiveMetadata#pushdownFilter` doesn't specify any predicate when constructing Constraint object. How about I add a TODO to add logic to specify the predicate?

```
        HivePartitionResult hivePartitionResult = partitionManager.getPartitions(metastore, tableHandle, new Constraint<>(entireColumnDomain), session);
```

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -1723,6 +1740,90 @@ public boolean supportsMetadataDelete(ConnectorSession session, ConnectorTableHa
     }
 
+    @Override
+    public boolean isPushdownFilterSupported(ConnectorSession session, ConnectorTableHandle tableHandle)
```
Can we add a unit tests or integration tests for this code path?

mbasmanova replied:
@highker This path is not fully implemented yet. Only the planner portion is available. Hence, an end-to-end queries don't work yet.

mbasmanova commented:
@highker 

> Can we have test coverage on this change? Rather than putting the test in TestPickTableLayout (which only aims to test a single rule), can we have the test in TestLogicalPlanner or any class inheriting BasePlanTest? For such case, we can run the entire planner.

It would be quite helpful to have a planner test for filter pushdown. The existing testing frameworks appear to be insufficient though. TestLogicalPlanner only works with TPCH connector which doesn't support filter pushdown. In particular, plan tests rely on LocalQueryRunner#createPlan method that returns a Plan object. All of Hive tests use DistributedQueryRunner which doesn't have createPlan method. DistributedQueryRunner can only be used to fetch a textual representation of the plan by running EXPLAIN queries. I'm going take a look to see if there is a way to implement createPlan that returns Plan object with (Hive)DistributedQueryRunner. In the meantime, if you have some other ideas about how to write such tests, let me know.

mbasmanova commented:
@highker I figured that I can add `assertPlan` method to `AbstractTestQueryFramework` and that will allow me to write planner tests using Hive connector (similar to TestLogicalPlanner). I'm going to add such tests and update the PR.

```

    public Plan assertPlan(String query, PlanMatchPattern pattern)
    {
        QueryExplainer explainer = getQueryExplainer();
        return transaction(queryRunner.getTransactionManager(), queryRunner.getAccessControl())
                .singleStatement()
                .execute(queryRunner.getDefaultSession(), session -> {
                    Plan actualPlan = explainer.getLogicalPlan(session, sqlParser.createStatement(query, createParsingOptions(session)), emptyList(), WarningCollector.NOOP);
                    PlanAssert.assertPlan(session, queryRunner.getMetadata(), queryRunner.getStatsCalculator(), actualPlan, pattern);
                    return null;
                });
    }
```

wenleix commented:
This PR is bundled with https://github.com/prestodb/presto/pull/12729 to support filter pushdown, and will be superseded once Connector can participate in planning, (see https://github.com/prestodb/presto/issues/12546 for details) right?

mbasmanova commented:
@wenleix The SPI changes in the PR will be superseded by the `Connector can participate in planning` things eventually.

mbasmanova commented:
@highker @hellium01 James, Yi, thank you for reviewing this PR. I made the following changes to address your feedback:

- Added TestHiveLogicPlanner test to verify planner part of the pushdown with Hive connector.
- Replaced TranslateVariableNamesVisitor with RowExpressionNodeInliner#replaceExpression
- Added a TODO about pruning partitions using on non-tuple-domain predicates

It feel like adding ColumnReferenceExpression is not strictly necessary for this PR. It is also not clear how exactly it will be used when connector will be working with the entire plan. Hence, I'd rather not include this in this PR.

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java:
```diff
@@ -250,9 +278,48 @@ public static PlanNode pushPredicateIntoTableScan(
+                    WarningCollector.NOOP,
+                    false);
+
+            BiMap<VariableReferenceExpression, VariableReferenceExpression> symbolToColumnMapping = node.getAssignments().entrySet().stream()
```
Curious: Why do we need `BiMap` here? Looks like `RowExpressionNodeInliner.replaceExpression` requires an `Map` as last parameter?

mbasmanova replied:
@wenleix The inverted mapping is used to convert unenforcedConstraint into an expression that is used to create a FilterNode on top of the TableScanNode. The call to `.inverse()` got lost in the last round of changes and the test didn't catch it because Hive connector never returns unenforcedConstraint that's not TRUE. I added the missing call.

```
            RowExpression unenforcedConstraint = pushdownFilterResult.getUnenforcedConstraint();
            if (!TRUE.equals(unenforcedConstraint)) {
                return new FilterNode(idAllocator.getNextId(), tableScan, replaceExpression(unenforcedConstraint, symbolToColumnMapping.inverse()));
            }
```

mbasmanova commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java:
```diff
@@ -0,0 +1,18 @@
+ */
+package com.facebook.presto.hive;
+
+public class TestHiveLogicalPlanner
```
@tdcmeehan I fixed this.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/metadata/PushdownFilterResult.java:
```diff
@@ -0,0 +1,40 @@
+public class PushdownFilterResult
+{
+    private final TableLayout layout;
+    private final RowExpression unenforcedConstraint;
```
@wenleix Sure. I'll rename it.

mbasmanova commented:
@wenleix 

> Or, is the idea here that for a connector supports pushdownFilter, this pushdownFilter supersedes getLayout for consistency purpose?

That's the plan: have the engine use either pushdownFilter or getTableLayouts consistently based on isPushdownFilterSupported. In this PR I replaced getTableLayouts with pushdownFilter in PickTableLayout. Subsequent PRs will be replacing getTableLayouts calls in other places.

To clarify, even when filter = TRUE, getTableLayouts cannot be used in place of pushdownFilter because it doesn't provide the "current" state (ConnectorTableLayoutHandle) and therefore the connector cannot know if a filter was pushed down before.

mbasmanova commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -1736,18 +1846,46 @@ public boolean supportsMetadataDelete(ConnectorSession session, ConnectorTableHa
+
+    private boolean isPushdownFilterEnabled(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        boolean pushdownFilterEnabled = HiveSessionProperties.isPushdownFilterEnabled(session);
```
I don't think I can do that. It will clash with this method.

mbasmanova commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveSplit.java:
```diff
@@ -102,7 +107,9 @@ public HiveSplit(
         this.tableBucketNumber = tableBucketNumber;
         this.forceLocalScheduling = forceLocalScheduling;
-        this.effectivePredicate = effectivePredicate;
+        this.domainPredicate = requireNonNull(domainPredicate, "domainPredicate is null");
```
Will do.

mbasmanova commented:
@tdcmeehan @highker Added the cool new `@Experimenal` annotation to methods in ConnectorMetadata.
